Maps Object

KISS opened the pull request:
There are two map objects: 

1. `Map--multi-src` accepts multiple geojson files and renders a toggle link for each source. When you select a link, the corresponding layer is displayed on the map. Only one layer is visible at a time. If only one geojson file is passed to this map, the map will render as expected (excluding the toggle link). 

<img width="558" alt="Screen Shot 2019-05-30 at 9 50 42 AM" src="https://user-images.githubusercontent.com/182383/58637469-851b4780-82c0-11e9-8067-57c678093a8d.png">

2. `Map--single-src` accepts only one geojson file. It expects a column (`legendColumn`) to be defined (in our case, "boro") in order to generate a legend containing all unique column values. When you select an item from the legend, all related shapes (in our case, "zipcode") are highlighted in the map.

<img width="546" alt="Screen Shot 2019-05-30 at 9 50 48 AM" src="https://user-images.githubusercontent.com/182383/58637473-88aece80-82c0-11e9-9475-46bd4decfa43.png">

In both maps, you can interact directly with a shape on the map. Doing so will highlight the selected shape and display a popup containing its `filterBy` value.

#Ask:
Would appreciate feedback on naming and ways to simplify the code since there is repetition. Perhaps moving `map.setLayoutProperty();` and `map.setFilter()` into a separate function (ex: `updateFilter()`) since they're used together frequently.

devowhippit commented:
@KISS Are the Multi and Single source files (src/objects/maps/Map--multi-src.js) needed in order for the map to function? It doesn't appear as though a user could configure them.

If you were to combine the files, you could move your method functions and assign them to the a class at the bottom of the page. I've found this to be the best way to store methods in a class that contains a Vue application because it retains the scope of the Vue application. An example model for this would be the Vue Applications (PEU and Programs Filter) on ACCESS; https://github.com/CityOfNewYork/ACCESS-NYC/blob/master/wp-content/themes/access/src/js/modules/wp-archive-vue.js

Basically the class would look more like this;

```
// map-vue.js

class Map {
  constructor(Vue, settings) {
    this._default = {
       methods: {
         someMethod: Map.someMethod
       }
     }

    this._settings = Object.assign({}, this._default, settings); // This is overly simplified, it wouldn't work because the .assign method does not perform deep recursion.

    return new Vue(settings);
  }
}

// Assign methods here.
Map.someMethod = function() {
  ... method does something...
}

export default Map;
```

By doing this, you can have some default methods to go along with your Map function but allow the user to pass along a customized Vue instance and attach any components they want to go with it. Below would be an example of a user integrating the above script in their own app.

```
import Vue from 'vue';
import Map from 'map-vue.js';
import SomeComponent from 'someComponent';

Vue.component('c-some-component', SomeComponent);

new Map(Vue, {
  methods: {
    customUserMethod: function() {
      ... custom method does something else ...
    }
  };
```

If you were to use this approach you could create one file with all of your methods in the same place. You could then pass a parameter to the Map object telling it to use 'single' or 'multi' methods.

KISS commented:
@devowhippit @kimpenguin 

Made the changes we talked about: 
1. merged vue files into one with conditional checks, 
2. moved js files to src/js,
3. updated "source" -> "layer"

I'm not sure if #2 was implemented the way you intended, so let me know if it requires another refactor. 

cc/ @abourne1 heads up (kim mentioned you're using the component), the functionality is the same but the way it's initialized is different and requires a `mapType` (`single` or `multi`) to be set in the config object, ping me if you have questions.

devowhippit commented on src/objects/map/map.vue:
```diff
@@ -0,0 +1,442 @@
+          ['#2F334F', '#ACAEB9'],
+          // blue, orange
+          ['#3155A6', '#F9A137']
+        ];
```
@KISS I think this is not a blocker but was curious if you thought there was any way we could externalize the color settings... I think it's ok if it's hard-coded but preferably we can manage it from the config/variables.js file. I suppose we could pass it through the properties.

KISS replied:
@devowhippit yeah, I considered having the colors set on the `layer` object but wanted to make sure there was proper color contrasting and that the colors matched those set for the library. figured hard-coding the values was simplest, since the layer color has to be set as a property when it's added to Mapbox.

would creating a similar attribute to `colorCombinations` in `variables.js` and then importing `variables.js` into `Maps.js`, so it can be used to set a `color` property on the layer, work better?

devowhippit replied:
@KISS if that is possible it sounds like the right path to me!

KISS replied:
@devowhippit I removed the hard coded values. I couldn't import `variables.js` directly because it has a `package.json` dependency so I created a `colors.js` config instead. I moved `colors` and `colorCombinations` into `colors.js` to keep everything in one place and require `colors.js` in `variables.js`. 

I also updated the repo so we can use markdown and updated the docs. To view the docs you'll need to add the `.html` ext: http://localhost:3000/map.html cc/ @kimpenguin

devowhippit commented:
@KISS I've just noticed when I stand up the application locally, the base layer map doesn't appear and there appears to be an issue with one of the requests to mapbox; 
![image](https://user-images.githubusercontent.com/3826974/59056184-a1415a80-8865-11e9-9006-baa33cc0ed3b.png)

Is this a keys issue? Perhaps documentation could address it?

KISS commented:
@devowhippit yeah, that's a key issue. I mention it in the docs (first para re: `token`) but will add the error message so it's clearer. (Mapbox uses "token" instead of "key" in their documentation)

kimpenguin commented:
Question. Should the `src/map.data.js` be kept with the object in `objects/map`?

KISS commented:
@kimpenguin I moved it out because it's not required for the component when being utilized by someone else but not tied to the change. cc/ @devowhippit thoughts?

devowhippit commented:
@kimpenguin @KISS I have kept .data.js files in similarly constructed components before. See the Filter Component: https://github.com/CityOfNewYork/ACCESS-NYC-PATTERNS/tree/master/src/components/filter

Haven't done a great job of documenting that you need to provide data to the component for it to render: https://cityofnewyork.github.io/ACCESS-NYC-PATTERNS-DEVELOPMENT/filter

KISS commented:
@devowhippit @kimpenguin - sounds good, I'll move it back :)

kimpenguin commented:
@KISS I made a couple edits to the markdown. LMK if I can push it into this branch :)

KISS commented:
@kimpenguin yeah, sounds good. go ahead :)

KISS commented:
@devowhippit merged `map-demo` but removed the new `.vue` and `.slm` files because I couldn't get them to render properly. need to revisit when the template structure is updated to match ACCESS.

cc/ @kimpenguin  

please confirm if this is ready to merge
